lib/diagnostics_channel.js

```diff
@@ -136,10 +138,136 @@ function hasSubscribers(name) {
   return channel.hasSubscribers;
 }
 
+class TracingChannel {
+  constructor(name) {
+    this.name = name;
+    this.channels = {
+      start: channel(`${name}.start`),
+      end: channel(`${name}.end`),
+      asyncEnd: channel(`${name}.asyncEnd`),
+      error: channel(`${name}.error`)
+    };
+  }
+
+  get hasSubscribers() {
+    const { channels } = this;
+    for (const key in channels) {
+      if (channels[key].hasSubscribers) return true;
+    }
+    return false;
+  }
+
+  subscribe(handlers) {
+    let subscribed = true;
+    for (const key in handlers) {
+      if (!subscribe(`${this.name}.${key}`, handlers[key])) {
+        subscribed = false;
+      }
+    }
+    return subscribed;
+  }
+
+  unsubscribe(handlers) {
+    let unsubscribed = true;
+    for (const key in handlers) {
+      if (!unsubscribe(`${this.name}.${key}`, handlers[key])) {
+        unsubscribed = false;
+      }
+    }
+    return unsubscribed;
+  }
+
+  traceSync(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, error } = this.channels;
+    start.publish(ctx);
+    try {
+      const result = ReflectApply(fn, thisArg, args);
+      ctx.result = result;
+      return result;
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  tracePromise(fn, ctx = {}, thisArg, ...args) {
+    const { asyncEnd, start, end, error } = this.channels;
+    start.publish(ctx);
+
+    const reject = (err) => {
+      ctx.error = err;
+      error.publish(ctx);
+      asyncEnd.publish(ctx);
+      throw err;
+    };
+
+    const resolve = (result) => {
+      ctx.result = result;
+      asyncEnd.publish(ctx);
+      return result;
+    };
+
+    try {
+      const promise = ReflectApply(fn, thisArg, args);
+      return PromisePrototypeThen(promise, resolve, reject);
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  traceCallback(fn, position = 0, ctx = {}, thisArg, ...args) {
+    const { start, end, asyncEnd, error } = this.channels;
+    start.publish(ctx);
+
+    function wrap(fn) {
+      return function wrappedCallback (err, res) {
+        if (err) {
+          ctx.error = err;
+          error.publish(ctx);
+        } else {
+          ctx.result = res;
+        }
+
+        asyncEnd.publish(ctx);
+        if (fn) {
+          return ReflectApply(fn, this, arguments);
+        }
+      }
+    }
```

Comment:
I would recommend we avoid wrapping altogether if there are no subscribers.

Reply:
Done. :)